### ISSUE TITLE
Shared features not working with --csoaa_ldf flag

### ISSUE THREAD

**Issue by bassmang** [VowpalWabbit/vowpal_wabbit#3088]
#### Describe the bug
Using the `--csoaa_ldf m` flag on a dataset with shared features produces buggy output.

#### To Reproduce
1. Used this specific command line: `./vw --csoaa_ldf m ex.txt`
2. Used `ex.txt`:
```
shared | g
1:0 | Real_f Auto_f RealAuto_f
2:1 | Toy_f  Auto_f ToyAuto_f
3:1 | Real_f Air_f  RealAir_f
4:1 | Toy_f  Air_f  ToyAir_f
```

#### Expected behavior
Num weight bits = 18                                                                                                                                                                                                                                                                                                                                      learning rate = 0.5                                                                                                                                                                                                                                                                                                                                       initial_t = 0                                                                                                                                                                                                                                                                                                                                             power_t = 0.5                                                                                                                                                                                                                                                                                                                                             using no cache                                                                                                                                                                                                                                                                                                                                            Reading datafile = ex                                                                                                                                                                                                                                                                                                                                     num sources = 1                                                                                                                                                                                                                                                                                                                                           average  since         example        example  current  current  current                                                                                                                                                                                                                                                                                  loss     last          counter         weight    label  predict features                                                                                                                                                                                                                                                                                  0.000000 0.000000            1            1.0    known        1       16                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            finished run                                                                                                                                                                                                                                                                                                                                              number of examples = 1                                                                                                                                                                                                                                                                                                                                    weighted example sum = 1.000000                                                                                                                                                                                                                                                                                                                           weighted label sum = 0.000000                                                                                                                                                                                                                                                                                                                             average loss = 0.000000                                                                                                                                                                                                                                                                                                                                   total feature number = 16   


#### Observed Behavior
Num weight bits = 18                                                                                                                                                                                                                                                                                                                                      learning rate = 0.5                                                                                                                                                                                                                                                                                                                                       initial_t = 0                                                                                                                                                                                                                                                                                                                                             power_t = 0.5                                                                                                                                                                                                                                                                                                                                             using no cache                                                                                                                                                                                                                                                                                                                                            Reading datafile = ex                                                                                                                                                                                                                                                                                                                                     num sources = 1                                                                                                                                                                                                                                                                                                                                           Enabled reductions: gd, scorer, csoaa_ldf, shared_feature_merger                                                                                                                                                                                                                                                                                          average  since         example        example  current  current  current                                                                                                                                                                                                                                                                                  loss     last          counter         weight    label  predict features                                                                                                                                                                                                                                                                                  [warning] NAN prediction in example 1, forcing 0                                                                                                                                                                                                                                                                                                          [warning] NAN prediction in example 1, forcing 0                                                                                                                                                                                                                                                                                                          [warning] NAN prediction in example 1, forcing 0                                                                                                                                                                                                                                                                                                          [warning] NAN prediction in example 1, forcing 0                                                                                                                                                                                                                                                                                                          -340282346638528859811704183484516925440.000000 -340282346638528859811704183484516925440.000000            1            1.0    known        0       18                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              finished run                                                                                                                                                                                                                                                                                                                                              number of examples = 1                                                                                                                                                                                                                                                                                                                                    weighted example sum = 1.000000                                                                                                                                                                                                                                                                                                                           weighted label sum = 0.000000                                                                                                                                                                                                                                                                                                                             average loss = -340282346638528859811704183484516925440.000000                                                                                                                                                                                                                                                                                            total feature number = 18     

#### Environment
Buggy behavior was in master branch, correct behavior was in old version.
Linux command line environment
